File: Graphs/Java/AdjacencyList/IDBased/PQueue.java
import java.io.*;
import java.util.Scanner;

class PQueue<T1,T2>{

    int size;
    int num_vd;
    Vertex<T1,T2> vd[];

    @SuppressWarnings("unchecked")
    public PQueue(int size){
        this.size = size;
        this.vd = (Vertex<T1,T2>[])new Vertex[size]; 
        this.num_vd = 0;

        for(int i=0;i<size;i++){
            vd[i] = new Vertex<>(-1,null,null);
            vd[i] = null;
        }

    }

    public static<T1,T2> void printQ(PQueue<T1,T2> q){
        
        System.out.println("Your stored edge weights are: ");

        for(int i=0;i<q.vd.length && q.vd[i] != null;i++)
        System.out.print(q.vd[i].edge_weight+" ");

        System.out.println("\n");
    }

    @SuppressWarnings("unchecked")
    public static<T1,T2> void dsizePQueue(PQueue<T1,T2> q){

        PQueue<T1,T2> temp = new PQueue<>(2*q.size);
        temp.num_vd = q.num_vd;

        for(int i=0;i<q.vd.length && q.vd[i]!=null;i++)
        temp.vd[i] = q.vd[i];
        
    
        q.vd = temp.vd;
        q.size = temp.size;
       
    }


    public static<T1,T2> Vertex<T1,T2> peek(PQueue<T1,T2> q){

        if(q.vd[0] == null){
             System.out.println("Your queue is empty");
             return null;
        }else return q.vd[0];
    } 


    public static<T1,T2> void insert(PQueue<T1,T2> q,Vertex<T1,T2> v){
        
        if(q.num_vd >= q.vd.length){
            dsizePQueue(q);
            q.vd[q.num_vd] = v;
            q.num_vd++;
        }
        else{
            q.vd[q.num_vd] = v;
            q.num_vd++;
        }

        bubbleUP(q,q.num_vd-1);
    }

    public static<T1,T2> void bubbleUP(PQueue<T1,T2> q,int last_element_index){

        if(last_element_index > 0){

            boolean b = false;
            int k = last_element_index;
            int parent_index = (k-1)/2;

            while(parent_index >= 0 && !b)
            {
            if(q.vd[parent_index].edge_weight > q.vd[k].edge_weight ){
                Vertex<T1,T2> temp = q.vd[k];
                q.vd[k] = q.vd[parent_index];
                q.vd[parent_index] = temp;
                k = parent_index;

                if(k-1 < 0)
                b = true;
                else parent_index = (k-1)/2;
            }else b = true;
          }
        }
    }

    public static<T1,T2> Vertex<T1,T2> pop(PQueue<T1,T2> q){
        
        Vertex<T1,T2> res = q.vd[0];
        
        if(q.num_vd > 0){
        if(q.vd[1] == null){
            q.vd[0] = null;
            q.num_vd--;
        }else {
              q.vd[0] = q.vd[q.num_vd-1];
              q.vd[q.num_vd-1] = null;
              q.num_vd--;
              bubbleDown(q);
        }
      }
        return res;
    }

    public static<T1,T2> void bubbleDown(PQueue<T1,T2> q){
        
        
        boolean b = false;
        int k = 0;

        while(!b){

                if(2*k+1 >= q.num_vd  && 2*k+2 >= q.num_vd)
                b = true;
                else{
                    if(2*k+2 < q.num_vd){
                        
                        if(q.vd[2*k+1].edge_weight <= q.vd[2*k+2].edge_weight){
                            
                            if(q.vd[k].edge_weight > q.vd[2*k+1].edge_weight){
                                Vertex<T1,T2> temp = q.vd[k];
                                q.vd[k] = q.vd[2*k+1];
                                q.vd[2*k+1] = temp;
                                k = 2*k+1;
                            }else b = true;
                        }else{
                            
                            if(q.vd[k].edge_weight > q.vd[2*k+2].edge_weight){
                                Vertex<T1,T2> temp = q.vd[k];
                                q.vd[k] = q.vd[2*k+2];
                                q.vd[2*k+2] = temp;
                                k = 2*k+2;
                            }else b = true;
                        }

                    }else{
                     
                            if(q.vd[k].edge_weight > q.vd[2*k+1].edge_weight){
                                Vertex<T1,T2> temp = q.vd[k];
                                q.vd[k] = q.vd[2*k+1];
                                q.vd[2*k+1] = temp;
                                k = 2*k+1;
                            }else b = true;
                    }
                }
        }
 
    }


public static void main(String args[]){

    Scanner sc = new Scanner(System.in);
    int s;
    int n;

    System.out.println("Give me the size of your queue");
    s = sc.nextInt();

    PQueue<String,String> qs = new PQueue<>(s);

    System.out.println("Enter the number of vertices would you like to insert");
    n = sc.nextInt();

    System.out.println("");

    int id = 0;
    while(n>0){
        System.out.println("Give me the label, the data, and the edge weight of your vertex "+id);

        if(sc.hasNextLine())
        sc.nextLine();

        String ltmp = sc.nextLine();
        String dtmp = sc.nextLine();
        int ew_tmp = sc.nextInt();
    

        Vertex<String,String> vtmp = new Vertex<>(id,ltmp,dtmp);
        vtmp.edge_weight = ew_tmp;

        insert(qs,vtmp);
        id++;
        n--;
    }

    System.out.println("\n");

    printQ(qs);

    System.out.println("After popping: \n");

    
    pop(qs);
    pop(qs);

    printQ(qs);

    System.out.println("\n");
    System.out.println("The number of vertices in you queue now is: \n"+qs.num_vd);
 
    
    }
  
}

File: Graphs/Python/AdjacencyList/IDBased/PQueue.py
import sys
import os
from Vertex import Vertex


class PQueue:

    def __init__(self,size):
        self.size = size
        self.num_vd = 0
        self.vd = size*[None]

        for i in range(size):
            self.vd[i] = Vertex(None,None,None)

    def printPQ(self):
        if self.num_vd == 0:
            print("The queue is empty")
        else:
            print("The vertices in queue are: ")
            i=0
            while i < self.num_vd and self.v[i] != None:
                print(self.vd[i].label,end=" ")
                i+=1
            
            print()

    def printWPQ(self):
        if self.num_vd == 0:
            print("The queue is empty")
        else:
            print("The edge weights are: ")
            i=0
            while i < self.num_vd and self.vd[i] != None:
                print(self.vd[i].edge_weight,end=" ")
                i+=1
                
            print()

    def dsizePQ(self):
        temp = PQueue(2*self.size)

        for i in range(len(self.vd)):
            temp.vd[i] = self.vd[i]
        
        self.vd = temp.vd
        self.size = temp.size 
        

    def insert(self,ve:Vertex):
        if self.num_vd >= len(self.vd):
            self.dsizePQ()
            self.vd[self.num_vd] = ve
            self.num_vd+=1
            self.bubbleUp()
        else:
            self.vd[self.num_vd] = ve
            self.num_vd+=1
            self.bubbleUp()
        

    def bubbleUp(self):
        b = False
        c_index = self.num_vd-1
        while not(b):
            p_index = (c_index-1)//2
            if c_index <= 0 or self.vd[c_index].edge_weight >= self.vd[p_index].edge_weight:
                b = True
            else:
                temp = self.vd[p_index]
                self.vd[p_index] = self.vd[c_index]
                self.vd[c_index] = temp
                c_index = p_index
                     
    def peek(self):
        if self.num_vd == 0:
            print("The queue is empty")
        else:
            return self.vd[0]

    def pop(self):
        if self.num_vd <= 1:
            temp = self.vd[0]
            self.vd[0] = None
            self.num_vd-=1
            return temp
        else:
            temp = self.vd[0]
            self.vd[0] = self.vd[self.num_vd-1]
            self.vd[self.num_vd-1] = None
            self.num_vd-=1
            self.bubbleDown()
            return temp

    def bubbleDown(self):
        b = False
        k = 0
        while not(b):
            lc_v = 2*k + 1
            rc_v =  2*k + 2

            if lc_v >= self.num_vd and rc_v >= self.num_vd:
                b = True
            else:
                if rc_v < self.num_vd:
                    if self.vd[lc_v].edge_weight <= self.vd[rc_v].edge_weight:
                        if self.vd[k].edge_weight > self.vd[lc_v].edge_weight:
                            temp = self.vd[k]
                            self.vd[k] = self.vd[lc_v]
                            self.vd[lc_v] = temp
                            k = lc_v
                        else:
                            b = True
                    else:
                        if self.vd[k].edge_weight > self.vd[rc_v].edge_weight:
                            temp = self.vd[k]
                            self.vd[k] = self.vd[rc_v]
                            self.vd[rc_v] = temp
                            k = rc_v
                        else:
                            b = True
                else:
                    if self.vd[k].edge_weight > self.vd[lc_v].edge_weight:
                        temp = self.vd[k]
                        self.vd[k] = self.vd[lc_v]
                        self.vd[lc_v] = temp
                        k = lc_v
                    else:
                        b = True
    
  


        

"""Testing"""

"""
print("Give me the size of your queue \n")
N = int(input())

pq = PQueue(N)

print("Give me the number of your vertices that you would like to insert \n")
nv = int(input())

for i in range(nv):
    print("Give me the id,label, data and the edge weight of vertex ",i,"\n")
    v_id = int(input())
    v_label = input()
    v_data = input()
    v_weight = int(input())
    v_e = Vertex(v_id,v_label,v_data)
    v_e.edge_weight = v_weight
    pq.insert(v_e)

print()
pq.printWPQ()
print()

pq.pop()
pq.pop()
print("After popping twice \n")
pq.printWPQ()
"""